lib/Dialect/ESI/ESIServices.cpp

```diff
@@ -109,8 +113,133 @@ static LogicalResult instantiateCosimEndpointOps(ServiceImplementReqOp req,
   return success();
 }
 
+// Generator for "sv_mem" implementation type. Emits SV ops for an unpacked
+// array, hopefully inferred as a memory to the SV compiler.
+static LogicalResult
+instantiateSystemVerilogMemory(ServiceImplementReqOp req,
+                               ServiceDeclOpInterface decl) {
+  ImplicitLocOpBuilder b(req.getLoc(), req);
+  BackedgeBuilder bb(b, req.getLoc());
+
+  RandomAccessMemoryDeclOp ramDecl =
+      dyn_cast<RandomAccessMemoryDeclOp>(decl.getOperation());
+  if (!ramDecl)
+    return req.emitOpError("'sv_mem' implementation type can only be used to "
+                           "implement RandomAccessMemory declarations");
+
+  if (req.getNumOperands() != 2)
+    return req.emitOpError("Implementation requires clk and rst operands");
```

Comment:
ughh... can we build in dedicated clock/reset ports (possibly through tagging ports with attributes)?

Reply:
For standard implementations, that's not a bad idea. This also affects the cosim implementation and would involve a number of changes to the operation (attribute vs string), but probably a worthwhile change!

Issue https://github.com/llvm/circt/issues/3942